Add enviroment map resource loader for <model>.

amarchenko opened the pull request:
#### 99eaeb7156ef3d4877571e8641a3a2b49e1bb02e
<pre>
Add enviroment map resource loader for &lt;model&gt;.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281318">https://bugs.webkit.org/show_bug.cgi?id=281318</a>
<a href="https://rdar.apple.com/97206044">rdar://97206044</a>

Reviewed by Mike Wyrzykowski and Ada Chan.

Initial implementation of Image Based Lighting resource loading for &lt;model&gt;.
Similar to source loading path, uses CachedResourceLoader with
new FetchOptions::Destination to determine the type of resource loaded.

Canonical link: <a href="https://commits.webkit.org/285113@main">https://commits.webkit.org/285113@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/379d6f7f6ae7a1458dcaa03c6d40c75c1c8e3cd5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71589 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24363 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75701 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22794 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73704 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58803 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22614 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56545 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15018 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74655 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46272 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61665 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36994 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42935 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19131 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64839 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19495 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77419 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15823 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18673 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64257 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15866 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64253 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15816 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12400 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6039 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46802 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1581 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47873 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49157 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on LayoutTests/model-element/model-element-environment-map-ready.html:
```diff
@@ -0,0 +1,55 @@
+<!DOCTYPE html> <!-- webkit-test-runner [ ModelElementEnabled=true ModelProcessEnabled=true ] -->
+<meta charset="utf-8">
+<title>&lt;model> ready promise</title>
```
Nit: can update the title to mention "environment map"

amarchenko replied:
Done.

achan00 commented on LayoutTests/model-element/model-element-environment-map-ready.html:
```diff
@@ -0,0 +1,55 @@
+
+    const [model, source] = createModelWithAttributesAndSource(t, {environmentmap: blueResource }, "resources/cube.usdz");
+    assert_true(model.environmentMap.includes(blueResource));
+    await model.ready;
```
Did you mean to wait for the `environmentMapReady` promise here?

amarchenko replied:
Yep, fixed.

achan00 commented on LayoutTests/model-element/model-element-environment-map-ready.html:
```diff
@@ -0,0 +1,55 @@
+
+    model.environmentMap = redResource
+    assert_true(model.environmentMap.includes(redResource));
+    await model.ready;
```
Ditto.

amarchenko replied:
Done.

mwyrzykowski commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -3361,6 +3361,7 @@ enum class WebCore::FetchOptionsDestination : uint8_t {
     Audioworklet,
     Document,
     Embed,
+    Environmentmap,
```
this should probably be `EnvironmentMap` with a capital-M, I'm not sure why the other values in the enum aren't camel cased like the other enums in this file

amarchenko replied:
I'd keep it as is for consistency sake but I agree with you.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -231,40 +232,67 @@ void HTMLModelElement::didAttachRenderers()
+        resourceHandle = nullptr;
 
         if (CheckedPtr renderer = this->renderer())
             renderer->updateFromElement();
```
Do we really need to do this renderer update for the environment map case? I didn't think we need to.

I think it may be clearer to just have two helper methods to handle the load completion of model resource and the environment map resource respectively. This way this `invalidateResourceHandleAndUpdateRenderer` block wouldn't need to take in a resource handle either.

Then this method will call the correct helper method depending on the `resource` identity.

amarchenko replied:
Added two helper methods, one for model resource and one for environment map resources. Model's resource original code kept without changes. This method calls corresponding helpers.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -688,6 +728,73 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+    m_environmentMapResource->addClient(*this);
+}
+
+void HTMLModelElement::updateEnvironmentmap()
```
Nit: `updateEnvironmentMap`?

amarchenko replied:
Done.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -688,6 +728,73 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+    m_environmentMapReadyPromise = makeUniqueRef<EnvironmentMapPromise>();
+
+    if (m_environmentMapURL.isEmpty()) {
+        ActiveDOMObject::queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, Event::create(eventNames().errorEvent, Event::CanBubble::No, Event::IsCancelable::No));
```
I don't think we should fire the error event here - wouldn't it conflict with the error event from regular model resource loading?

amarchenko replied:
Removed events sending, let's leave it to the spec to decide when and what events we send. Also, empty URL is a valid value that indicates removal of IBL, added corresponding code for that.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -688,6 +728,73 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+
+    auto resource = document().protectedCachedResourceLoader()->requestEnvironmentMapResource(WTFMove(request));
+    if (!resource.has_value()) {
+        ActiveDOMObject::queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, Event::create(eventNames().errorEvent, Event::CanBubble::No, Event::IsCancelable::No));
```
Ditto.

amarchenko replied:
Removed.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -688,6 +728,73 @@ void HTMLModelElement::setCurrentTime(double currentTime)
+
+    if (m_environmentMapURL.isEmpty()) {
+        ActiveDOMObject::queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, Event::create(eventNames().errorEvent, Event::CanBubble::No, Event::IsCancelable::No));
+        return;
```
We also need to handle removing the environment map (if it's set from a non-empty value to an empty one, or set from a non-empty IBL to a non-existent one). We'd want to end up calling `removeIBL` on the `WKSRKEntity` in that case.

amarchenko replied:
Added code to handle resource removal.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.h:
```diff
@@ -155,6 +157,8 @@ class ModelProcessModelPlayerProxy final
     bool m_loop { false };
     double m_playbackRate { 1.0 };
+
+    RefPtr<WebCore::SharedBuffer> m_environmentMapData;
```
Since `m_environmentMapData` is something that you reset after you've passed it over to `WKSRKEntity` but not represent the current state of the environment map, I wonder if it should be named to `m_pendingEnvironmentMapData` to be clearer.

amarchenko replied:
Changed.

achan00 replied:
Looks like the `m_environmentMapData` in `HTMLModelElement` got renamed instead of this one in `ModelProcessModelPlayerProxy`. We can just fix this in a later patch in this area.

achan00 commented on Source/WebKit/ModelProcess/cocoa/ModelProcessModelPlayerProxy.mm:
```diff
@@ -571,6 +573,24 @@ static CGFloat effectivePointsPerMeter(CALayer *caLayer)
+{
+    m_environmentMapData = WTFMove(data);
+    if (m_modelRKEntity)
+        applyEnvironmentMapDataAndRelease();
```
We need to handle the removal case as mentioned earlier.

amarchenko replied:
Added.

achan00 reviewed (approved):
Thanks for addressing the feedback!

amarchenko committed:
Add enviroment map resource loader for <model>.
https://bugs.webkit.org/show_bug.cgi?id=281318
rdar://97206044

Reviewed by Mike Wyrzykowski and Ada Chan.

Initial implementation of Image Based Lighting resource loading for <model>.
Similar to source loading path, uses CachedResourceLoader with
new FetchOptions::Destination to determine the type of resource loaded.

Canonical link: https://commits.webkit.org/285113@main
